### PR TITLE
code changes for adding support for form parameter matching

### LINKED ISSUE #383: Please add Form parameter matching ( Not query parameter matching)
Hi,

I am using WireMock 2.0.10-beta version. I didn't see any API for Form parameter matching. Not query parameter matching.

Proposed API

withFormParam("p1", equalTo("v1"));

OR

withRequestParam("p1", equalTo("v1"));  //  Work for both GET & Non-GET methods

Thank you!

### PR DESCRIPTION
**Pull request by kapishmalik**
Resolves #383 

Code changes for adding support for form parameter matching. 

Added tests for mentioned scenarios:

- Correct serialisation and deserialisation of the matcher.
- Parameter keys that contain characters that would be URL encoded, in particular array-style keys like address[postcode]
- Expected parameter values containing characters that would be URL encoded.
- Diff report

### CODE REVIEW

File: src/test/java/com/github/tomakehurst/wiremock/matching/RequestPatternTest.java
Comment (oleg-nenashev) [inline src/test/java/com/github/tomakehurst/wiremock/matching/RequestPatternTest.java:38]: Would be better to keep formatting of the original tests even if it violates the standard. Reviiews with `?w=1`  is a solution, of course

File: src/main/java/com/github/tomakehurst/wiremock/http/Request.java
Comment (oleg-nenashev) [inline src/main/java/com/github/tomakehurst/wiremock/http/Request.java:63]: Without default implementation, it breaks binary compatibility

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Thanks for the pull request! The binary compatibility issue in the interface needs to be fixed before this pull request can be merged.

**Comment by oleg-nenashev** [comment]
FWIW, there is https://github.com/wiremock/wiremock/issues/2153 for providing clear separation between really public APIs and not so ones

File: src/test/java/com/github/tomakehurst/wiremock/matching/RequestPatternTest.java
Comment (kapishmalik) [reply src/test/java/com/github/tomakehurst/wiremock/matching/RequestPatternTest.java:38]: If I keep it same, then spotless check will fail.

File: src/main/java/com/github/tomakehurst/wiremock/http/Request.java
Comment (kapishmalik) [reply src/main/java/com/github/tomakehurst/wiremock/http/Request.java]: incorporated

**Comment by kapishmalik** [comment]
incorporated review comment changes

**Review by oleg-nenashev — APPROVED** [review]
LGTM

File: src/main/java/com/github/tomakehurst/wiremock/http/Request.java
Comment (tomakehurst) [inline src/main/java/com/github/tomakehurst/wiremock/http/Request.java:63]: Can we ditch these defaults and ensure we've implemented these everywhere instead?

**Comment by tomakehurst** [comment]
Apart from the couple of minor comments above, looks good. Happy to merge this once these are resolved and Oleg's comment on the doc PR is also resolved.

**Comment by tomakehurst** [comment]
Apart from the couple of minor comments above, looks good. Happy to merge this once these are resolved and Oleg's comment on the doc PR is also resolved.

File: src/main/java/com/github/tomakehurst/wiremock/http/Request.java
Comment (kapishmalik) [reply src/main/java/com/github/tomakehurst/wiremock/http/Request.java]: Done. We have implemented these everywhere. Oleg suggested adding this as it might break the implementation for those who might be using this interface as this is part of the library and exposed.

File: src/test/java/com/github/tomakehurst/wiremock/matching/RequestPatternTest.java
Comment (kapishmalik) [reply src/test/java/com/github/tomakehurst/wiremock/matching/RequestPatternTest.java:1]: I just added my change but spotless check started failing. It was very minor test so I just removed it for now. Let me know if you have any work around to fix this.

File: src/main/java/com/github/tomakehurst/wiremock/http/Request.java
Comment (kapishmalik) [reply src/main/java/com/github/tomakehurst/wiremock/http/Request.java]: I feel that it is very unlikely that anyone using it as we are using this interface as base for our receiving requests, matching, and stubbing it wherever required. Consumers of Wiremock may not be using it for any of these purposes. But it is always better to add a safety net.

**Comment by tomakehurst** [comment]
Looks like you've deleted your changes to `RequestPatternTest` entirely now?

We need some serialisation/deserialisation tests, which I think belong in there.

**Comment by kapishmalik** [comment]
> Looks like you've deleted your changes to `RequestPatternTest` entirely now?
> 
> We need some serialisation/deserialisation tests, which I think belong in there.

as discussed, first applied formatting changes followed by added new UTs.

File: src/test/java/com/github/tomakehurst/wiremock/matching/RequestPatternTest.java
Comment (tomakehurst) [inline src/test/java/com/github/tomakehurst/wiremock/matching/RequestPatternTest.java:322]: Please can we also have a test case showing that serialisation works correctly?

Comment (kapishmalik) [reply src/test/java/com/github/tomakehurst/wiremock/matching/RequestPatternTest.java:322]: sure, incorporated

Comment (kapishmalik) [reply src/test/java/com/github/tomakehurst/wiremock/matching/RequestPatternTest.java:359]: I believe you mean deserialization. I have now both tests serialization and deserialization.

**Comment by tomakehurst** [comment]
Great stuff, thanks!

**Pull request merged**
